src/psyclone/psyir/nodes/array_mixin.py

```diff
@@ -320,6 +318,30 @@ def indices(self):
                     f"expression but found '{type(child).__name__}'")
         return self.children
 
+    @property
+    def shape(self):
+        '''
+        :returns: the shape of the array access represented by this node.
+        :rtype: List[:py:class:`psyclone.psyir.nodes.DataNode`]
+
+        '''
+        shape = []
+        for idx_expr in self.indices:
+            if isinstance(idx_expr, Range):
```

Comment:
They can return array slices and, surprisingly (to me), gfortran is quite happy to have something like `a(b(1:4))`.

    logical, dimension(0:111,0:111,0:111) :: stencil
    integer, dimension(4) :: indices = (/2, 3, 4, 5/) 
  
    stencil(indices(2:3), 1, 1) = .true.
    write(*,*) stencil(indices(:), 1, 1)
    write(*,*) size(stencil(indices(:), 1, 1), 1)

prints:

    F T T F
            4

This requires some thought!

Reply:
Since the indexing into the array `a` is now indirect, we don't know which elements of `a` will be accessed. All we can say for sure is that there is a 1D array constructed from elements of `a` in any order. I'm not sure whether we can call this a slice as it won't necessarily consist of contiguous members of `a`.